File: .java/Kira Loves Palindromes.java
import java.util.*;

class Solution {

  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    String s = in.next();
    Trie root = new Trie();
    long ans = 0;
    int len = s.length();
    for (int i = 0; i < len; i++) {
      root.addWord(s.substring(i), i);
    }
    s = new StringBuilder(s).reverse().toString();
    for (int i = 0; i < len; i++) {
      ans += root.searchWord(s.substring(i), len - 1 - i);
    //   System.out.println(ans);
    }
    System.out.println(ans);

  }

  static private boolean isPalindrome(char[] chArr, int l, int r) {
    while (l < r && chArr[l] == chArr[r]) {
      l++;
      r--;
    }
    return l >= r;
  }

  static class Trie {
    node root;

    Trie() {
      this.root = new node();
    }

    public void addWord(String word, int start) {
      node cur = root;
      char[] chArr = word.toCharArray();
      for (int i = 0; i < chArr.length; i++) {
        char ch = chArr[i];
        if (cur.child[ch - 'a'] == null) {
          cur.child[ch - 'a'] = new node();
        }
        cur = cur.child[ch - 'a'];
        for (int j = i + 1; j < chArr.length; j++) {
          if (isPalindrome(chArr, i + 1, j)) {
            cur.pal.add(start + j);
          }
        }
        cur.list.add(start + i);
      }
    }

    public int searchWord(String word, int end) {
      int ans = 0;
      node cur = root;
      char[] chArr = word.toCharArray();
      for (int i = 0; i < chArr.length; i++) {
        char ch = chArr[i];
        if (cur.child[ch - 'a'] == null) {
          return ans;
        }
        cur = cur.child[ch - 'a'];
        for (int j = i; j < chArr.length; j++) {
          if (isPalindrome(chArr, i + 1, j)) {
            int index = Collections.binarySearch(cur.list, end - j);
            index = index >= 0 ? index : -index - 1;
            ans += index;
            // System.out.println("** " + ans);
          }
        }
        for (int p : cur.pal) {
            if (p < end - i) {
                ans++;
            }
        }
      }
     return ans;
    }
  }

  static class node {
    List<Integer> list; // list of end indexs of substrings which end at this node
    List<Integer> pal; // list of end indexs of substrings whose suffix is palindrome
    node[] child;

    node() {
      list = new ArrayList<>();
      pal = new ArrayList<>();
      child = new node[26];
    }
  }
}

File: Kira Loves Palindromes.cpp
#include <bits/stdc++.h>
using namespace std;

struct node {
  vector<int> v1;
  vector<int> v2;
  node* child[26];
};

void addWord(node* root, const char* word, int start);
int searchWord(node* root, const char* word, int end);
const int LENMAX = 1001;
bool p[LENMAX][LENMAX];
int main(int argc, char const* argv[]) {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  char* s = new char[LENMAX];
  cin >> s;
  int len = strlen(s);
  for (int d = 0; d < len; d++) {
    for (int i = 0; i + d < len; i++) {
      int j = i + d;
      p[i][j] = s[i] == s[j] && (i + 1 > j - 1 || p[i + 1][j - 1]);
    }
  }
  node* trie = new node();
  long ans = 0;
  for (int i = 0; i < len; i++) {
    addWord(trie, s, i);
  }
  for (int i = len - 1; i >= 0; i--) {
    ans += searchWord(trie, s, i);
  }
  cout << ans << "\n";
  return 0;
}

void addWord(node* root, const char* word, int start) {
  node* cur = root;
  for (int i = start; word[i]; i++) {
    char ch = word[i];
    if (cur->child[ch - 'a'] == NULL) {
      cur->child[ch - 'a'] = new node();
    }
    cur = cur->child[ch - 'a'];
    for (int j = i + 1; word[j]; j++) {
      if (p[i + 1][j]) {
        cur->v2.push_back(j);
      }
    }
    cur->v1.push_back(i);
  }
};
int searchWord(node* root, const char* word, int end) {
  int ans = 0;
  node* cur = root;
  for (int i = end; i >= 0; i--) {
    char ch = word[i];
    if (cur->child[ch - 'a'] == NULL) {
      return ans;
    }
    cur = cur->child[ch - 'a'];
    for (int j = i; j >= 0; j--) {
      if (j == i || p[j][i - 1]) {
        ans += lower_bound(cur->v1.begin(), cur->v1.end(), j) - cur->v1.begin();
      }
    }
    for (int p : cur->v2) {
      if (p < i) {
        ans++;
      }
    }
  }
  return ans;
};